Add material design features to Input

abarth opened the pull request:
This pull request has two commits (which might be easier to review individually):
1. Add material design features to Input
   
   Properly support labels, hints, icons, and custom typography.
2. Improve the factoring between Input and RawEditableLine
   
   RawEditableLine is now responsible for the scrolling behavior, which removes the need for callbacks between RawEditableLine and Input. It also fixes a bug whereby the whole Input widget (including its icon) would scroll when the text got long.

@eseidelGoogle @jason-simmons

abarth committed:
Add material design features to Input

Properly support labels, hints, icons, and custom typography.

abarth committed:
Improve the factoring between Input and RawEditableLine

RawEditableLine is now responsible for the scrolling behavior, which
removes the need for callbacks between RawEditableLine and Input. It
also fixes a bug whereby the whole Input widget (including its icon)
would scroll when the text got long.

abarth committed:
Add implicit animations to Input

Now the label animates from its inline position to above the text and
the focus highlight grows into place.

abarth commented:
3. Added some implicit animations.

eseidelGoogle commented on examples/address_book/lib/main.dart:
```diff
@@ -4,79 +4,41 @@
+          new Input(key: _kNameKey, icon: 'social/person', labelText: 'Name', style: Typography.black.display1),
+          new Input(key: _kPhoneKey, icon: 'communication/phone', hintText: 'Phone'),
+          new Input(key: _kEmailKey, icon: 'communication/email', hintText: 'Email'),
+          new Input(key: _kAddressKey, icon: 'maps/place', hintText: 'Address'),
```
Intersting.  I didn't know that icons were part of input.  Makes some sense I guess.  Certainly for a material.Input.

abarth replied:
Input is part of the material package.  We name it something else if that's confusing.

eseidelGoogle commented on packages/flutter/lib/src/material/input.dart:
```diff
@@ -95,49 +128,124 @@ class InputState extends ScrollableState<Input> {
+    TextStyle textStyle = config.style ?? themeData.text.subhead;
+    Color focusHighlightColor = themeData.accentColor;
+    if (themeData.primarySwatch != null)
+      focusHighlightColor = focused ? themeData.primarySwatch[400] : themeData.hintColor;
```
We're probably going to want a named constant for 400 (not necessarily this patch).  No clue why 400 is right instead of 500 here.

eseidelGoogle commented on packages/flutter/lib/src/material/input.dart:
```diff
@@ -95,49 +128,124 @@ class InputState extends ScrollableState<Input> {
+    Color focusHighlightColor = themeData.accentColor;
+    if (themeData.primarySwatch != null)
+      focusHighlightColor = focused ? themeData.primarySwatch[400] : themeData.hintColor;
+    double topPadding = config.isDense ? 12.0 : 16.0;
```
I guess all tehse are specified in teh spec.

abarth replied:
Yeah, these values just come straight from the spec.  They aren't re-used anywhere as far as I can tell.

eseidelGoogle commented on packages/flutter/lib/src/material/input.dart:
```diff
@@ -95,49 +128,124 @@ class InputState extends ScrollableState<Input> {
     ));
 
+    if (config.errorText != null && !config.isDense) {
+      TextStyle errorStyle = themeData.text.caption.copyWith(color: Colors.red[700]);
```
You uses red[700] twice.  presumably as the error color.

abarth replied:
We could put that in ThemeData.  I'm not sure if it shows up elsewhere.  The spec doesn't say, but using digital color meter told me it was ren[700].

abarth replied:
Done.

eseidelGoogle commented on packages/flutter/lib/src/material/input.dart:
```diff
@@ -95,49 +128,124 @@ class InputState extends ScrollableState<Input> {
+        children: [
+          new Container(
+            margin: new EdgeDims.only(right: 16.0, top: iconTop),
+            width: config.isDense ? 40.0 : 48.0,
```
Crazy all teh custom dimentions for the isDense stuff. :)

abarth replied:
Yeah, the dense stuff seems like a wart on the design.  We could leave it out of the implementation, but I'm sure someone will want to use it.

eseidelGoogle commented:
This is a huge improvement, thanks.  lgtm regardless of my nits above.

abarth committed:
Add errorColor to ThemeData

That way folks can customize it if they want (and it makes the code a
bit more self-documenting).

Hixie commented on packages/flutter/lib/src/material/theme_data.dart:
```diff
@@ -263,12 +273,15 @@ class ThemeData {
+        text,
+        primaryTextTheme,
+        primaryIconTheme
+      )
```
!! we exceeded the 20 values that hashValues() supports?! Blimey. Filed https://github.com/flutter/flutter/issues/1356.

Hixie commented on packages/flutter/lib/src/widgets/editable.dart:
```diff
@@ -212,14 +274,17 @@ class RawEditableTextState extends State<RawEditableLine> {
-      hideText: config.hideText,
-      onContentSizeChanged: config.onContentSizeChanged,
-      scrollOffset: config.scrollOffset
+    return new SizeObserver(
```
nooooooooooooooooooooooooooooooooooooooooooooo

abarth replied:
I just moved it from one place to another.  It's the same problem that all the scrollables have about wanting to know how big their outsides are for the scrolling physics.
